cmd/kubeadm/app/constants/constants.go

```diff
@@ -209,6 +209,9 @@ const (
 	// KubeadmConfigConfigMap specifies in what ConfigMap in the kube-system namespace the `kubeadm init` configuration should be stored
 	KubeadmConfigConfigMap = "kubeadm-config"
 
+	// KubeadmConfigLockConfigMap specifies what ConfigMap in the kube-system namespace should be used for locking when modifying the KubeadmConfigConfigMap shared resource
+	KubeadmConfigLockConfigMap = "kubeadm-config-lock"
```

Comment:
my understanding was that we can use the kubeadm-config map, but make sure that we don't stomp Data from the lock and don't stomp the annotation when updating Data from our existing methods.

this is based on the comments by @detiber in the original ticket and my later investigations.

Reply:
I think it is definitely possible to do locking on the `kubeadm-config` map with a few caveats.

1) All access to `kubeadm-config` configmap needs to be wrapped with a lock
2) All updates to the `kubeadm-config` configmap should probably be done through a PATCH to avoid overwriting the annotations
3) There should probably be a method to 'unlock' the configmap instead of waiting for the timeout to expire

I think the biggest challenge is that the current LeaderElection locks are not really designed to be used around mutating objects and there isn't native support for ensuring that the lock annotations are persisted across updates.

It would be great if there was native support in the apiserver for locking a resource and limiting access to that resource while it's locked.

Reply:
> 3. There should probably be a method to 'unlock' the configmap instead of waiting for the timeout to expire

Assuming the client will be able to contact the apiserver, the leader election config allows for a `ReleaseOnCancel`, that will call to `release()` function, removing the leadership annotation. If something goes wrong on the client side and `release()` is not called, or the request does not reach the apiserver for any reason, then the lock would be held until the timeout happens, what I believe is a fair compromise. Is this what you expect?